### PR TITLE
blockscout integration

### LINKED ISSUE #29: Enable updating optimism repositories
For now the OP Labs repositories (optimism monorepo, op-geth) repo are only pulled once and set to the required version, after which we assume they're set to the version required by the script.

We should at least perform this check explicitly, and then offer to update (pull + switch to right commit) if that's not the case. Maybe even offer to reset or stash changes in the repo if there are any.

### LINKED ISSUE #45: Enable starting a block explorer
The codebase now clones the Blockscout code and does work on it, but the end result right now is it doesn't work.

This was a hackathon effort (h/t @chee-chyuan), so it's not super clear to me what exactly breaks.

### LINKED ISSUE #62: Clean need remove explorer service containers
We may need removing explorer service containers when #61 merged? @norswap

### PR DESCRIPTION
**Pull request by GrapeBaBa**
fixes #45 
fixes #62 
fixes #29

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -106,23 +106,28 @@ def setup_op_geth_repo():
...
+    git_tag = "b6136977051730a7acd47b35b6bdcf9a74e39be9"
 
     if os.path.isfile("blockscout"):
         raise Exception("Error: 'blockscout' exists as a file and not a directory.")
     elif not os.path.exists("blockscout"):
         print("Cloning the blockscout repository. This may take a while...")
         lib.clone_repo(github_url, "clone the blockscout repository")
 
-        lib.run("checkout stable version", f"git checkout --detach {git_tag}",
+        lib.run("checkout optimism stable version", f"git checkout --detach {git_tag}",
```
Comment (norswap) [inline setup.py:117]: let's say "checkout blockscout's optmism stable version"

What makes this the "optimism stable version" btw? Could we use a tag instead? Could we include the branch name here in a comment?

File: setup.py [hunk 2]
```diff
@@ -106,23 +106,28 @@ def setup_op_geth_repo():
...
                 cwd="blockscout")
 
         # TODO make this not replace multiple times if run multiple times
+        anchor_line = "image: blockscout/blockscout:${DOCKER_TAG:-latest}"
```
Comment (norswap) [inline setup.py:121]: It looks like this isn't used anymore and can be removed?

Comment (norswap) [inline setup.py:121]: Could you also take care of the TODO above, if it isn't already the case?

**Review by norswap — CHANGES_REQUESTED** [review]
Have you been able to run the explorer and check that everything is functional?

I was able to run it during the hackathon and get to the page but it didn't really work (it was just syncing... I think?).

At minimum we should be able to see the system transactions for every block!

File: setup.py (hunk 2 above)
Comment (GrapeBaBa) [reply setup.py:121]: it still used, the L126 replace it with a specific optimism image which i tested works.

Comment (GrapeBaBa) [reply setup.py:121]: it still used, the L126 replace it with a specific optimism image which i tested works.

Comment (GrapeBaBa) [reply setup.py:121]: i think i didn't try to fix the TODO case

File: setup.py (hunk 1 above)
Comment (GrapeBaBa) [reply setup.py:117]: right now i just use the a specific commit of optimism branch, they don't have a tag for the branch. It is indeed not stable version, however it is fair difficult to get a stable version for our optimism code.

**Comment by GrapeBaBa** [comment]
> Have you been able to run the explorer and check that everything is functional?
> 
> I was able to run it during the hackathon and get to the page but it didn't really work (it was just syncing... I think?).
> 
> At minimum we should be able to see the system transactions for every block!

have you checkout this PR to try. 

> Have you been able to run the explorer and check that everything is functional?
> 
> I was able to run it during the hackathon and get to the page but it didn't really work (it was just syncing... I think?).
> 
> At minimum we should be able to see the system transactions for every block!

changed frontend image back, it seems there is a new image published by blockscout. 

```sh
./rollop setup --yes  
./rollop --clean devnet --explorer  
```

open localhost:80 should works.

File: setup.py (hunk 1 above)
Comment (norswap) [reply setup.py:117]: Gotcha, let's just add a comment that says `// commit in the "optimism" branch" (use the actual branch name if it's not exactly "optimism")

**Comment by norswap** [comment]
> have you checkout this PR to try.

No, I was just reviewing very quickly, but I'll definitely try it out later. Just wanted to confirm :)

**Comment by GrapeBaBa** [comment]
> > have you checkout this PR to try.
> 
> No, I was just reviewing very quickly, but I'll definitely try it out later. Just wanted to confirm :)

I tested several times with the last two commit image tags, it could shows blocks and transactions which i assume it works basiclly

**Comment by norswap** [comment]
Hey, I'm sorry, sort of lost track of this.

I'm trying it right now and it does not seem to work for me. I can access blockscout, but it keeps saying " 0% Blocks Indexed - We're indexing this chain right now. Some of the counts may be inaccurate." in a banner at the top, which is what it was doing before as well. This persists even if letting it run for 10+ minutes, and this is of course a chain that has been created merely a few seconds before blockscout is launched.

The blockscout launch seems to be reporting a lot of errors, though I'm not sure what they indicate:

```
blockscout               | ** (ArgumentError) Could not convert receipt to elixir
blockscout               | 
blockscout               | Receipt:
blockscout               |   %{"blockHash" => "0xf6fa36eed182319d352c3551a432a72c175c07b03296b5d0b4c5c48c11b3f9c6", "blockNumber" => "0x2a", "contractAddress" => nil, "cumulativeGasUsed" => "0xcce5", "depositNonce" => "0x29", "effectiveGasPrice" => "0x0", "from" => "0xdeaddeaddeaddeaddeaddeaddeaddeaddead0001", "gas" => 1000000, "gasUsed" => "0xcce5", "logs" => [], "logsBloom" => "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000", "status" => "0x1", "to" => "0x4200000000000000000000000000000000000015", "transactionHash" => "0x78f80e3f408a20cd4c3794cd3dcdda5d43fb83fd42d3ab89876e1a9fce7dd4cd", "transactionIndex" => "0x0", "type" => "0x7e"}
blockscout               | 
blockscout               | Errors:
blockscout               |   {:unknown_key, %{key: "depositNonce", value: "0x29"}}
blockscout               | 
blockscout               |     (ethereum_jsonrpc 5.1.0) lib/ethereum_jsonrpc/receipt.ex:236: EthereumJSONRPC.Receipt.ok!/2
blockscout               |     (elixir 1.13.1) lib/enum.ex:1593: Enum."-map/2-lists^map/1-0-"/2
blockscout               |     (ethereum_jsonrpc 5.1.0) lib/ethereum_jsonrpc/receipts.ex:138: EthereumJSONRPC.Receipts.fetch/2
blockscout               |     (elixir 1.13.1) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
blockscout               |     (elixir 1.13.1) lib/task/supervised.ex:34: Task.Supervised.reply/4
blockscout               |     (stdlib 3.17) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```

What I did was pull this branch then:

```
./rollop setup
./rollop --clean devnet --explorer
```

Did you use a different procedure?

**Comment by GrapeBaBa** [comment]
> Hey, I'm sorry, sort of lost track of this.
> 
> I'm trying it right now and it does not seem to work for me. I can access blockscout, but it keeps saying " 0% Blocks Indexed - We're indexing this chain right now. Some of the counts may be inaccurate." in a banner at the top, which is what it was doing before as well. This persists even if letting it run for 10+ minutes, and this is of course a chain that has been created merely a few seconds before blockscout is launched.
> 
> The blockscout launch seems to be reporting a lot of errors, though I'm not sure what they indicate:
> 
> ```
> blockscout               | ** (ArgumentError) Could not convert receipt to elixir
> blockscout               | 
> blockscout               | Receipt:
> blockscout               |   %{"blockHash" => "0xf6fa36eed182319d352c3551a432a72c175c07b03296b5d0b4c5c48c11b3f9c6", "blockNumber" => "0x2a", "contractAddress" => nil, "cumulativeGasUsed" => "0xcce5", "depositNonce" => "0x29", "effectiveGasPrice" => "0x0", "from" => "0xdeaddeaddeaddeaddeaddeaddeaddeaddead0001", "gas" => 1000000, "gasUsed" => "0xcce5", "logs" => [], "logsBloom" => "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000", "status" => "0x1", "to" => "0x4200000000000000000000000000000000000015", "transactionHash" => "0x78f80e3f408a20cd4c3794cd3dcdda5d43fb83fd42d3ab89876e1a9fce7dd4cd", "transactionIndex" => "0x0", "type" => "0x7e"}
> blockscout               | 
> blockscout               | Errors:
> blockscout               |   {:unknown_key, %{key: "depositNonce", value: "0x29"}}
> blockscout               | 
> blockscout               |     (ethereum_jsonrpc 5.1.0) lib/ethereum_jsonrpc/receipt.ex:236: EthereumJSONRPC.Receipt.ok!/2
> blockscout               |     (elixir 1.13.1) lib/enum.ex:1593: Enum."-map/2-lists^map/1-0-"/2
> blockscout               |     (ethereum_jsonrpc 5.1.0) lib/ethereum_jsonrpc/receipts.ex:138: EthereumJSONRPC.Receipts.fetch/2
> blockscout               |     (elixir 1.13.1) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
> blockscout               |     (elixir 1.13.1) lib/task/supervised.ex:34: Task.Supervised.reply/4
> blockscout               |     (stdlib 3.17) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
> ```
> 
> What I did was pull this branch then:
> 
> ```
> ./rollop setup
> ./rollop --clean devnet --explorer
> ```
> 
> Did you use a different procedure?

Did you remove the previous blockscout repo manually?

**Comment by norswap** [comment]
So I retried after delete the directory, but I'm running into mac-specific issues. If trying with the exact code, it complains that it can find the image, but it is `linux/amd64` instead of the requested `linux/arm64` — it's indeed what can be found [here](https://hub.docker.com/r/blockscout/blockscout-optimism/tags?page=1&name=3c6).

If I remove the modification and try to pull the `linux/amd64` image (trying to make it work via emulation) then blockscout crashes on startup.

This appears multiple times, that might be the issue (and not mac specific):

```
db                              | 2023-11-11 19:00:58.259 UTC [234] FATAL:  password authentication failed for user "postgres"
db                              | 2023-11-11 19:00:58.259 UTC [234] DETAIL:  Role "postgres" does not exist.
db                              | 	Connection matched pg_hba.conf line 100: "host all all all scram-sha-256"
```

But it's annoying given that postgres is launched in Docker as well (as explained [here](https://docs.blockscout.com/for-developers/information-and-settings/docker-integration-local-use-only)).

Re-running, I get other db issues, but no longer the password ones.

The problem, however, is that the backend is crashing with `init terminating in do_boot`.

Next thing: I tried [5.3.0-postrelease-b6136977](https://hub.docker.com/layers/blockscout/blockscout-optimism/5.3.0-postrelease-b6136977/images/sha256-af653ff0313fc51430c3d0d6b4460f30de111ef19fbf728ff7771912bb0afeee?context=explore)

This fails because there are too many connections to postgres (both on the arm64 and amd64 variants):

```
backend                         | 2023-11-11T20:58:22.579 application=db_connection [error] Postgrex.Protocol (#PID<0.4401.0>) failed to connect: ** (Postgrex.Error) FATAL 53300 (too_many_connections) sorry, too many clients already
```

Next step, I'm removing `DOCKER_TAG=1.55 FRONTEND_DOCKER_TAG=1.55`. But this causes it to download a bunch of things and my connection is super slow right now. Will continue to look into this later.

**Comment by GrapeBaBa** [comment]
> So I retried after delete the directory, but I'm running into mac-specific issues. If trying with the exact code, it complains that it can find the image, but it is `linux/amd64` instead of the requested `linux/arm64` — it's indeed what can be found [here](https://hub.docker.com/r/blockscout/blockscout-optimism/tags?page=1&name=3c6).
> 
> If I remove the modification and try to pull the `linux/amd64` image (trying to make it work via emulation) then blockscout crashes on startup.
> 
> This appears multiple times, that might be the issue (and not mac specific):
> 
> ```
> db                              | 2023-11-11 19:00:58.259 UTC [234] FATAL:  password authentication failed for user "postgres"
> db                              | 2023-11-11 19:00:58.259 UTC [234] DETAIL:  Role "postgres" does not exist.
> db                              | 	Connection matched pg_hba.conf line 100: "host all all all scram-sha-256"
> ```
> 
> But it's annoying given that postgres is launched in Docker as well (as explained [here](https://docs.blockscout.com/for-developers/information-and-settings/docker-integration-local-use-only)).
> 
> Re-running, I get other db issues, but no longer the password ones.
> 
> The problem, however, is that the backend is crashing with `init terminating in do_boot`.
> 
> Next thing: I tried [5.3.0-postrelease-b6136977](https://hub.docker.com/layers/blockscout/blockscout-optimism/5.3.0-postrelease-b6136977/images/sha256-af653ff0313fc51430c3d0d6b4460f30de111ef19fbf728ff7771912bb0afeee?context=explore)
> 
> This fails because there are too many connections to postgres (both on the arm64 and amd64 variants):
> 
> ```
> backend                         | 2023-11-11T20:58:22.579 application=db_connection [error] Postgrex.Protocol (#PID<0.4401.0>) failed to connect: ** (Postgrex.Error) FATAL 53300 (too_many_connections) sorry, too many clients already
> ```
> 
> Next step, I'm removing `DOCKER_TAG=1.55 FRONTEND_DOCKER_TAG=1.55`. But this causes it to download a bunch of things and my connection is super slow right now. Will continue to look into this later.
> 
> EDIT: finished downloading, still experiencing the same error

I don't have an mac with arm, let me ask others to try it.

**Comment by norswap** [comment]
Now works. The indexing % is wonky, only goes down, but restarting the explorer (by running it separately via `./rollop explorer`, quitting and restarting) seems to bump it 100%, then it drops again.

We need to try this on Linux and see how it works there.

**Comment by norswap** [comment]
I also ended up updating optimism to 1.3.1 and op-geth (which I forgot in the update to 1.3.0, though somehow that didn't change anything??).

Also I fixed quite a few misc issues, the most annoying being that the most recent version of foundry causes the deploy script to fail. We now install a local known version of foundry to avoid this.

**Pull request merged**